doc/logging/persistent_logger/persistent_loglevel.md

```diff
@@ -0,0 +1,460 @@
+   { CONFIG_DB,           TABLE_NAME_SEPARATOR_VBAR  },
+   { PFC_WD_DB,           TABLE_NAME_SEPARATOR_COLON },
+   { FLEX_COUNTER_DB,     TABLE_NAME_SEPARATOR_COLON },
+   { STATE_DB,            TABLE_NAME_SEPARATOR_VBAR  },
+   { APPL_STATE_DB,       TABLE_NAME_SEPARATOR_COLON }
+};
+```
+
+ - In addition we will remove the LOGLEVEL DB connector from /sonic-swss/fdbsyncd/fdbsyncd.cpp that not in use.
+
+## 3.2 Persistent Logger flow
+
+- Each component has a singleton Logger object with a log level property and listener thread.
+- When a component writes a log message, the Logger writes the message only if the loglevel of the message is above the current loglevel property.
+- When the user wants to set a new log level to a component, he uses the "swssloglevel" script. The "swssloglevel" script sets the new verbosity to the CONFIG DB (database #4).
+- The CONFIG DB change triggers an event, which is caught by the listener thread.
+- The listener thread changes the loglevel property of the Logger.
+- The user can use the "config save" command to save the current loglevel and make it persistent. It will copy the CONFIG DB content to the config_db.json.
+
+In addition to the log level, the CONFIG DB contains the log output file. After "config save" the log output will be persistent to reboot too.
+
+
+
+
+
+
+
+![persistent logger flow](/doc/logging/persistent_logger/persistent-logger-flow.png)
+
+
+
+
+
+
+
+# 4 Flows
+
+## 4.1 System init flow
+
+When the system startup and the Database container initialize, and the config_db.json loads to CONFIG DB.
+ 
+  - There is a concern about the log level of messages written before the database container is initialized (if it exists). From my understanding, the RSYSLOG-CONFIG is up only after the DATABASE container, so there won't be logs before the config_db.json loads to the CONFIG DB. In case there are messages before the loading, the messages will be in the default log level.
+
+
+# 5 Cold and Fast Reboot Support
+  
+  The current implementation support cold and fast reboot. Since in the cold and fast reboot, the CONFIG DB content is deleted, the user needs to run "config save" to make the log level persistent to cold and fast reboot.
+
+# 6 Warm Reboot Support
+
+  With the current implementation, we don't flush the CONFIG DB before warm-reboot, which means that if the user configures some loglevel (for example, debug), after warm-reboot, the system startup with the same configurable loglevel (debug).
+
+## 6.1 Support warm upgrade
+
+ Since we are not flushing the LOGLEVEL DB when we perform a warm upgrade, we need to update the db migrator.
+  - Add LOGLEVEL DB connector in the db migrator.
+  - In case of a warm upgrade, move Logger tables from LOGLEVEL DB to CONFIG DB (with a minor change in the key) and delete LOGLEVEL DB.
+  - Exposing the "del" function from the swss-common/sonicv2-connector.h to the db migrator. 
+
+(Downgrade will not be supported).
```

Comment:
>Downgrade will not be supported [](http://example.com/codeflow?start=1&length=31)

@yxieca Can we skip downgrade in design?

Reply:
Good catch. I don't think the design has to work for downgrade. However, downgrade is an important operation. It is important to state what would be the behavior for downgrade.

Reply:
Thanks @yxieca! I am mostly worried the process:
1. running with an image with this feature
2. downgrade to older image without this feature. By default, the new ConfigDB table is kept and no processes consume it
3. upgrade again to an image with this feature: since the config_db.json is carried over to new image, will it be override by "Logger tables from LOGLEVEL DB"?

Reply:
In this case, after step 1: we will have the Logger tables in the CONFIG DB and the LOGLEVEL DB will be empty.
after step 2: the CONFIG DB will contain the old Logger tables and the LOGLEVEL DB will contain new Logger tables with default values.
after step 3: in case of warm-upgrade the content of the LOGLEVEL DB will replace the content from the CONFIG DB.

Reply:
That's nice design consideration. I am looking forward to have this in the design doc itself.

Reply:
I just updated it, thanks!